crates/iceberg/src/spec/transform.rs

```diff
@@ -261,6 +269,300 @@ impl Transform {
+        width: Option<u32>,
+    ) -> Result<Option<Predicate>> {
+        if let Some(boundary) = self.projected_boundary(op, datum)? {
+            let transformed = func.transform_literal_result(&boundary)?;
+            let adjusted = self.adjust_projection(op, datum, &transformed);
+            let op = self.projected_operator(op, datum, width);
+
+            if let Some(op) = op {
+                let predicate = match adjusted {
+                    None => Predicate::Binary(BinaryExpression::new(
+                        op,
+                        Reference::new(name),
+                        transformed,
+                    )),
+                    Some(AdjustedProjection::Single(d)) => {
+                        Predicate::Binary(BinaryExpression::new(op, Reference::new(name), d))
+                    }
+                    Some(AdjustedProjection::Set(d)) => Predicate::Set(SetExpression::new(
+                        PredicateOperator::In,
+                        Reference::new(name),
+                        d,
+                    )),
+                };
+                return Ok(Some(predicate));
+            }
+        };
+
+        Ok(None)
+    }
+
+    /// Create a new `Datum` with adjusted projection boundary.
+    /// Returns `None` if `PredicateOperator` and `PrimitiveLiteral`
+    /// can not be projected
+    fn projected_boundary(&self, op: &PredicateOperator, datum: &Datum) -> Result<Option<Datum>> {
+        let literal = datum.literal();
+
+        let projected_boundary = match op {
+            PredicateOperator::LessThan => match literal {
+                PrimitiveLiteral::Int(v) => Some(Datum::int(v - 1)),
+                PrimitiveLiteral::Long(v) => Some(Datum::long(v - 1)),
+                PrimitiveLiteral::Decimal(v) => Some(Datum::decimal(v - 1)?),
+                PrimitiveLiteral::Date(v) => Some(Datum::date(v - 1)),
+                PrimitiveLiteral::Timestamp(v) => Some(Datum::timestamp_micros(v - 1)),
+                _ => Some(datum.to_owned()),
+            },
+            PredicateOperator::GreaterThan => match literal {
+                PrimitiveLiteral::Int(v) => Some(Datum::int(v + 1)),
+                PrimitiveLiteral::Long(v) => Some(Datum::long(v + 1)),
+                PrimitiveLiteral::Decimal(v) => Some(Datum::decimal(v + 1)?),
+                PrimitiveLiteral::Date(v) => Some(Datum::date(v + 1)),
+                PrimitiveLiteral::Timestamp(v) => Some(Datum::timestamp_micros(v + 1)),
+                _ => Some(datum.to_owned()),
+            },
+            PredicateOperator::Eq
+            | PredicateOperator::LessThanOrEq
+            | PredicateOperator::GreaterThanOrEq
+            | PredicateOperator::StartsWith
+            | PredicateOperator::NotStartsWith => Some(datum.to_owned()),
+            _ => None,
+        };
+
+        Ok(projected_boundary)
+    }
+
+    /// Create a new `PredicateOperator`, rewritten for projection
+    fn projected_operator(
```

Comment:
it applies to `Transform::Truncate` as well as the temporal transforms e.g. `Transform::Year` etc. I would leave the name as is since it applies to more than one transform?

Reply:
Reasonable to me.